### PR TITLE
feat: Externalize TLS cert for argocd-server

### LINKED ISSUE #5890: Separate TLS cert from admin password in argocd-secret
# Change
Split `argocd-secret` into opaque secret `argocd-secret` (admin password etc...) and TLS secret `argocd-cert`.

# Motivation
This allows an easier time configuring the certificates for argocd, for example with cert-manager. You can uncouple the TLS certificate rotation from argocd password.
You can also have an easier time with the various ingress setups in the documentation this way, for example with https://argoproj.github.io/argo-cd/operator-manual/ingress/#option-2-multiple-ingress-objects-and-hosts you end up with only one valid SAN.

### PR DESCRIPTION
**Pull request by jannfis**
Closes #5890

## Description

This PR adds support from using an _external_ secret for the TLS configuration of the Argo CD API Server (`argocd-server`). This allows for managing TLS configuration using third-party tools such as `cert-manager`, while still being able to properly have Argo CD self-manage using Git. This will also let users properly store their TLS configuration in secrets unmarshaled by encryption tools such as `SealedSecrets`, without having to store everything in that secret (which possibly gets overwritten by Argo CD Settings Manager).

Usage of `argocd-server-tls` secret is optional, and the change is implemented in a non-breaking way.

It will behave as follows:

* When a secret named `argocd-server-tls` exists in Argo CD's namespace, the TLS configuration will be read from its `tls.crt` and `tls.key` fields. Argo CD will ignore the fields `tls.crt` and `tls.key` in `argocd-secret` in this case, also during settings update, e.g. will not persist the key pair from the `argocd-server-tls` secret into `argocd-secret`.

* When the `argocd-server-tls` secret does not exist, current behaviour will not change, i.e. Argo CD will generate a self-signed keypair upon startup when `tls.key` and `tls.crt` do not exist in `argocd-secret` and persist this configuration in `argocd-secret`.

* Changes to `argocd-server-tls` secret will be picked up in the same way as the changes to `argocd-secret` secret are being picked up. There is no need  to restart `argocd-server` after changes to TLS configuration.

Signed-off-by: jannfis <jann@mistrust.net>

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

File: util/settings/settings.go [hunk 1]
```diff
@@ -992,6 +1008,28 @@ func updateSettingsFromSecret(settings *ArgoCDSettings, argoCDSecret *apiv1.Secr
...
+// externalServerTLSCertificate will try and load a TLS certificate from an
+// external secret, instead of tls.crt and tls.key in argocd-secret. If both
+// return values are nil, no external secret has been configured.
+func (mgr *SettingsManager) externalServerTLSCertificate() (*tls.Certificate, error) {
+	var cert tls.Certificate
+	secret, err := mgr.clientset.CoreV1().Secrets(mgr.namespace).Get(mgr.ctx, externalServerTLSSecretName, metav1.GetOptions{})
+	if err != nil {
+		if apierr.IsNotFound(err) {
+			return nil, nil
+		}
```
Comment (jannfis) [inline util/settings/settings.go:1020]: I was thinking whether to just return `err` here and have the caller check for `IsNotFound(err)` to decide. But I'm unsure.

**Review by alexmt — CHANGES_REQUESTED** [review]
Awesome change @jannfis ! I agree that the use case is valid and we need to make changes to support it.

Before we merge the PR I want to discuss an ehnancement:  Instead of hard-coding secret names, can we let the user define sensitive data in any secret in Argo CD namespace (labeled with `app.kubernetes.io/part-of: argocd`) ?
The implementation would be a little more complex: 
* To retrieve secret key it would have to check the list of secrets instead of one secret
* To update the key it would need to find if there is a secret that already has the key and update it, otherwise fallback to `argocd-secret` Secret
* Finally handle duplicated keys  secrets (pick first and log warning)

The advantage is that this approach would solve more use cases https://github.com/argoproj/argo-cd/issues/4052, https://github.com/argoproj/argo-cd/pull/4342

**Comment by jannfis** [comment]
Hey @alexmt, thank you for having a look.

I'm not entirely sure whether I understood you correctly, but if are suggesting to look for a secret that has `tls.crt` and `tls.key` keys and is labeled with `app.kubernetes.io/part-of: argocd`, I think we will have a hard time finding the right one to use for the following reasons:

* We may have an arbitrary number of secrets in the namespace which have `tls.crt` and `tls.key` in `data`, because there is more than one TLS configuration we need to manage (for example, TLS secret for repo server introduced in https://github.com/argoproj/argo-cd/pull/5764)
* Changing the name of the keys to something else than `tls.key` and `tls.crt` is not really desirable, because it's a de-facto Kubernetes standard (see https://kubernetes.io/docs/concepts/configuration/secret/#tls-secrets) for specifying TLS information

I understand that we might want to get rid of hard-coded secret names for good reasons. My first thought was that we could use an additional label on the secret, such as  `app.kubernetes.io/component` to specify which component the secret belongs to, so that the secret would be selected if

* Label `app.kubernetes.io/part-of` matches `argocd`
* Label `app.kubernetes.io/component` matches e.g. `argocd-server` (for API server endpoint) or `argocd-repo-server` (for the repository server endpoint)
* The `type` field matches `kubernetes.io/tls`

However, in case of the repository server's TLS secret, we also need to volume-mount it to the pods (at least the repo server pod), so we need to know its name because afaik Kubernetes doesn't let you use another selector for volume mounts here (according to https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-files-from-a-pod, which states `Modify your Pod definition to add a volume under .spec.volumes[]. Name the volume anything, and have a .spec.volumes[].secret.secretName field equal to the name of the Secret object.`. Also, third party tools that create such secrets may or may not support additional information to be propagated to the secret (such as additional labels).

So my personal opinion on this is, that we should use the name as convention for TLS secrets.

**Review by alexmt — APPROVED** [review]
LGTM

Thank you for the explanation @jannfis . I was focused on finding a way to avoid hard-coded names for secrets that don't need it. I agree that TLS secret is an example of a secret that needs it. I will focus on merging https://github.com/argoproj/argo-cd/pull/4342 instead.

**Pull request merged**